doc/developer/overview.md

```diff
@@ -0,0 +1,116 @@
+The optimizations are intended foremost to not change the "semantics" of the query (more on that in a moment).
+The optimizations are otherwise intended to improve the performance, resource use, and general tractability of dataflow execution.
+
+### Intended semantics
+
+Each of the `MirRelationExpr` variants are data-parallel operators that act on collections of input data.
+Each variant represents (or intends to represent) a deterministic function of its input data.
+This function maps collections of non-negative records to an output collection and error collection.
+Errors are a natural and expected result of queries with flaws; they are not meant to represent bugs in Materialize.
+Error collections are accumulated across the operators of a query; operators do not act on them.
+
+The main caveat here is that following SQL, we may arbitrarily move `MirScalarExpr` expressions throughout a query plan.
+The order of predicates within a `MirRelationExpr::Filter` or the placement of erroring expressions after a `Filter` or `Join` may still result in errors.
+The `MirScalarExpr::If` expression can be used to guard the evaluation of expressions and prevent errors.
+This movement of expressions should not change the correct output of a computation, but may change the content of the sidecar error collection.
+
+### Caveats
+
+The optimization layer relies on several properties of the expressions that it optimizes.
+The *types* of the expressions, including the data types themselves, non-nullability, and unique key structure, should be correct as stated: optimizations may take advantage of this information and change the semantics (or error, crash) on collections that violate them.
+The input collections are presumed to not contain records that accumulate to negative totals.
+The scalar expressions (`MirScalarExpr`) are presumed to be deterministic; the optimizer may change the structure of expressions to remove, reduce, or reuse apparently equivalent expressions.
+
+## Dataflow execution
+
+The differential dataflow engine communicates and operates on "updates" to collections of data.
+These updates are structured as `(data, time, diff)` triples, where
+
+1. `data` is usually a `Row` of data,
+2. `time` is usually the timestamp type of the coordinator,
+3. `diff` is usually a signed integer.
+
+Differential dataflow creates dataflow topologies of deterministic operators, each of which responds to updates in their inputs with the corresponding updates to their outputs: how should the output *change* in order to correct it to correspond to the operators logic applied to the changed input.
+
+These topologies are static, determined at dataflow construction time.
+It is important that the operators be deterministic, as determinism is the basis for "incremental recomputation" as a function of changed inputs.
```

Comment:
Could you add some examples regarding what "deterministic" means in practice? There are probably examples that come to your mind of how creators of operators have violated this determinism principle.

Reply:
The simplest example I can think of is 
`create materialized view bar as select rand(a) from foo`. If `foo` had `n` rows, but you then deleted a row, then differential would not be able to properly retract data because it is unable to able to figure out what value it originally produced for that row.

A more complex example is that within a dataflow, filters must run a fixed order. Suppose you have `create materialized view bar as select * from foo where a + 1/a > 0 and a != 0`. If `a + 1/a > 0` runs first, the dataflow will return an error if a row has `a = 0`. If `a != 0` runs first, then there will never be any error. If a dataflow chooses to run `a + 1/a > 0` first when a row with `a=0` is added but runs `a != 0 ` first when that row is retracted, the dataflow may get into in the erroneous state where it is returning an error even though there are no rows where `a=0`. 

That said, if you had,
`create materialized view bar as select * from foo where a + 1/a > 0 and a != 0`
`create materialized view bar2 as select * from foo where a + 1/a > 0 and a != 0`
it is perfectly acceptable if the dataflow for `bar` chooses to always run `a + 1/a > 0` first but the dataflow for bar2 chooses to always run `a != 0` first.